Empty rpath-link path prevents linking libstdc++ with clang++ under Linux

Hi,

My meson build fails to link with clang++ on Linux because clang++ doesn't link in libstdc++.  This works with autotools, and it works with meson and gcc.  The reason seems to be that (for both g++ and clang++) the link command ends with `-Wl,-rpath-link,`.  This doesn't bother gcc, but it does bother clang++, probably because `-Wl,-rpath-link,` absorbs the next argument to the linker, which is probably `-lstdc++` or similar.

If I remove this or change it to  `-Wl,-rpath-link,yo` then it works.  Also, if I change
```
            args += ['-Wl,-rpath-link,' + lpaths]
```
to
```
            if lpaths != '':
                args += ['-Wl,-rpath-link,' + lpaths]
```
in compilers/compilers.py  Does this sound like the right fix?  Maybe the test should be for something containing only whitespace.

-BenRI

OK, I got a reduced testcase:
[testcase.tar.gz](https://github.com/mesonbuild/meson/files/1584438/testcase.tar.gz)
``` sh
% cd testcase
% CC=clang CXX=clang++ meson build
% ninja -C build
% ninja -C build
ninja: Entering directory `build'
[2/2] Linking target testcase.
FAILED: testcase 
clang++  -o testcase 'testcase@exe/testcase.cc.o' -Wl,--no-undefined -Wl,--as-needed -Wl,-rpath,XXXXXXXXXXXXXXXXXXXXXXXX -Wl,-rpath-link,  
testcase@exe/testcase.cc.o: In function `main':
/home/bredelings/testcase/build/../testcase.cc:6: undefined reference to `operator new(unsigned long)'
/home/bredelings/testcase/build/../testcase.cc:6: undefined reference to `std::allocator<char>::allocator()'
/home/bredelings/testcase/build/../testcase.cc:6: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, std::allocator<char> const&)'
/home/bredelings/testcase/build/../testcase.cc:6: undefined reference to `std::allocator<char>::~allocator()'
/home/bredelings/testcase/build/../testcase.cc:7: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'
/home/bredelings/testcase/build/../testcase.cc:7: undefined reference to `operator delete(void*)'
/home/bredelings/testcase/build/../testcase.cc:6: undefined reference to `std::allocator<char>::~allocator()'
/home/bredelings/testcase/build/../testcase.cc:6: undefined reference to `operator delete(void*)'
/home/bredelings/testcase/build/../testcase.cc:7: undefined reference to `operator delete(void*)'
testcase@exe/testcase.cc.o: In function `__cxx_global_var_init':
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/c++/7.2.0/iostream:74: undefined reference to `std::ios_base::Init::Init()'
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/c++/7.2.0/iostream:74: undefined reference to `std::ios_base::Init::~Init()'
testcase@exe/testcase.cc.o:(.eh_frame+0x47): undefined reference to `__gxx_personality_v0'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

--- a/mesonbuild/compilers/compilers.py
+++ b/mesonbuild/compilers/compilers.py
@@ -868,7 +868,7 @@ class Compiler:
             # This eats the next argument, which happens to be 'ldstdc++', causing link failures.
             # We can dodge this problem by not adding any rpath_paths if the argument is empty.
             if lpaths.strip() != '':
-                args += ['-Wl,-rpath-link,'+lpaths]
+                args += ['-Wl,-rpath-link,' + lpaths]
         return args
 
 

--- a/mesonbuild/compilers/compilers.py
+++ b/mesonbuild/compilers/compilers.py
@@ -863,7 +863,12 @@ class Compiler:
             # Not needed on Windows or other platforms that don't use RPATH
             # https://github.com/mesonbuild/meson/issues/1897
             lpaths = ':'.join([os.path.join(build_dir, p) for p in rpath_paths])
-            args += ['-Wl,-rpath-link,' + lpaths]
+
+            # clang expands '-Wl,rpath-link,' to ['-rpath-link'] instead of ['-rpath-link','']
+            # This eats the next argument, which happens to be 'ldstdc++', causing link failures.
+            # We can dodge this problem by not adding any rpath_paths if the argument is empty.
+            if lpaths.strip() != '':
+                args += ['-Wl,-rpath-link,'+lpaths]
         return args
 
 

--- a/run_unittests.py
+++ b/run_unittests.py
@@ -2288,11 +2288,18 @@ class LinuxlikeTests(BasePlatformTests):
         testdir = os.path.join(self.unit_test_dir, '11 build_rpath')
         self.init(testdir)
         self.build()
+        # C program RPATH
         build_rpath = get_rpath(os.path.join(self.builddir, 'prog'))
         self.assertEqual(build_rpath, '$ORIGIN/sub:/foo/bar')
         self.install()
         install_rpath = get_rpath(os.path.join(self.installdir, 'usr/bin/prog'))
         self.assertEqual(install_rpath, '/baz')
+        # C++ program RPATH
+        build_rpath = get_rpath(os.path.join(self.builddir, 'progcxx'))
+        self.assertEqual(build_rpath, '$ORIGIN/sub:/foo/bar')
+        self.install()
+        install_rpath = get_rpath(os.path.join(self.installdir, 'usr/bin/progcxx'))
+        self.assertEqual(install_rpath, 'baz')
 
     def test_pch_with_address_sanitizer(self):
         testdir = os.path.join(self.common_test_dir, '13 pch')

--- a/test cases/unit/11 build_rpath/meson.build
+++ b/test cases/unit/11 build_rpath/meson.build
@@ -1,9 +1,16 @@
-project('build rpath', 'c')
+project('build rpath', 'c', 'cpp')
 
 subdir('sub')
 executable('prog', 'prog.c',
   link_with : l,
   build_rpath : '/foo/bar',
   install_rpath : '/baz',
+  install : true,
+  )
+
+executable('progcxx', 'prog.cc',
+  link_with : l,
+  build_rpath : '/foo/bar',
+  install_rpath : 'baz',
   install : true,
   )

--- a/test cases/unit/11 build_rpath/prog.cc
+++ b/test cases/unit/11 build_rpath/prog.cc
@@ -1,0 +1,8 @@
+#include <string>
+#include <iostream>
+
+int main(int argc, char **argv) {
+    std::string* s = new std::string("Hello");
+    delete s;
+    return 0;
+}

